{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imposing Model Constraints \n",
    "\n",
    "Here we take measured gas exchange, metabolite and proteomics data in order to constraint the flux of the reaction in the model by [Arnold and Nikoloski, 2014](https://www.ncbi.nlm.nih.gov/pubmed/24808102).\n",
    "\n",
    "We build four models using the experimental data: \n",
    "- Wild-type 20 degrees,\n",
    "- Wild-type after 7 days in 4 degrees, \n",
    "- fum2 mutant 20 degrees, \n",
    "- fum2 mutant after 7 days in 4 degrees. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Loading Required Packages \n",
    "# import cobra\n",
    "# from cobra.flux_analysis import flux_variability_analysis\n",
    "# from cobra.flux_analysis.loopless import add_loopless, loopless_solution\n",
    "# from cobra.flux_analysis import pfba\n",
    "# import pandas as pd\n",
    "# import numpy as np\n",
    "# import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading the protein data \n",
    "df = pd.read_excel(io=\"ProteinConc.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the model \n",
    "%run SetUpModel.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions for running the models \n",
    "\n",
    "def ATPNADPH(atp,nadph):\n",
    "    '''Setting the ATP and NADPH flux. Lower bound plus one'''\n",
    "    NADPH = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "    ATPase = model.reactions.get_by_id('ATPase_h')\n",
    "    lohi = 0.1\n",
    "    NADPH.lower_bound = nadph - lohi\n",
    "    NADPH.upper_bound = nadph + lohi\n",
    "    ATPase.lower_bound = atp- lohi\n",
    "    ATPase.upper_bound = atp + lohi\n",
    "    print(NADPH)\n",
    "    print(NADPH.bounds)\n",
    "    print(ATPase)\n",
    "    print(ATPase.bounds)\n",
    "    print(\" \")\n",
    "    \n",
    "def setobjective(FunName,obj=\"Maximum\"):\n",
    "    '''Setting the objective function of the model to FunName'''\n",
    "    model.objective = model.reactions.get_by_id(FunName)\n",
    "    model.objective_direction = obj\n",
    "    print(model.objective)\n",
    "    print(\"Set as objective function. \\n\")\n",
    "    \n",
    "def constr(reacs,Dat1,Dat2,Dat3,Dat4):\n",
    "    '''Constraints model reactions for a specific genotype and temperature'''\n",
    "    for r in reacs:\n",
    "        upper_vals = []\n",
    "        lower_vals = []\n",
    "        for g in r.genes:\n",
    "            # Get index of protein in data file \n",
    "            ind = all_atg.index(g.id)\n",
    "            # Scaling the values for setting fluxes \n",
    "            vals = [i/4000. for i in Dat1[ind],Dat2[ind],Dat3[ind],Dat4[ind]]\n",
    "            val = np.mean(vals)\n",
    "            sterr = np.std(vals)/np.sqrt(len(vals))\n",
    "            upper = val + sterr\n",
    "            lower = val - sterr\n",
    "            upper_vals.append(upper)\n",
    "            lower_vals.append(lower)\n",
    "        # Constraint flux of forward reaction\n",
    "        r.upper_bound = round(sum(upper_vals),3)\n",
    "        # Constraint flux of lower reaction \n",
    "        if r.lower_bound < 0.0:\n",
    "            r.lower_bound = round(-sum(upper_vals),3)\n",
    "        # No lower bound b/c don't know localization or enzymes are used to in other reactions \n",
    "\n",
    "def TPTratio(TPT1,TPT2):\n",
    "    '''To set TPT1 to TPT2 flux ratios'''\n",
    "    r = model.reactions.get_by_id('Tr_TPT2')\n",
    "    r.lower_bound = 0.0\n",
    "    r.upper_bound = TPT2\n",
    "    r = model.reactions.get_by_id('Tr_TPT1')\n",
    "    r.lower_bound = 0.0\n",
    "    r.upper_bound = TPT1\n",
    "    #print(cobra.flux_analysis.flux_variability_analysis(model, [NADPH,ATPase]))\n",
    "    print(\" \")\n",
    "    #inds = (1,101,202,302,401,501,601,701,801,901)\n",
    "    \n",
    "def subsys(subsystem):\n",
    "    for r in model.reactions:\n",
    "        if r.subsystem == subsystem:\n",
    "            print(r.reaction)\n",
    "            \n",
    "def input_output_reacs(model):\n",
    "    \"Prints all input and output reactions for the given model\"\n",
    "    reacs = []\n",
    "    nums = []\n",
    "    counter = 0\n",
    "    for r in model.reactions:\n",
    "        coeffs = []\n",
    "        for i in r.metabolites:\n",
    "            c = r.get_coefficient(i.id)\n",
    "            coeffs.append(c)\n",
    "        if all(i > 0.0 for i in coeffs) or all(i < 0.0 for i in coeffs) :\n",
    "            reacs.append(r)\n",
    "            print(r)\n",
    "            nums.append(counter)\n",
    "        counter = counter + 1\n",
    "    return(reacs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of all possible proteins in the model\n",
    "gene_ids = []\n",
    "for g in model.genes:\n",
    "    gene_ids.append(g.id)\n",
    "# Get list of all measured proteins \n",
    "all_atg = []\n",
    "x=df[\"Accession\"]\n",
    "for atg in x.values:\n",
    "    all_atg.append(atg)\n",
    "# Comparing the measured protein concentrations to those which are in the model \n",
    "model_atg = [] # Total measured proteins in model\n",
    "model_atg_final = [] # Proteins with direct reaction constraint (i.e. single responsible protein)\n",
    "model_rxns_atg = [] # Model reactions with direct constraints (i.e. single responsible protein)\n",
    "model_rxns_atgs = [] # Model reactions with multiple possible constraints \n",
    "model_atgs_final = [] # Proteins with multiple possible constraints \n",
    "\n",
    "for atg in x.values: # Check if protein is in the model \n",
    "    if atg in gene_ids:\n",
    "        model_atg.append(atg)\n",
    "        # Only parameterize that reaction if that enzyme is the only one responsible for that reaction\n",
    "        for r in model.genes.get_by_id(atg).reactions:\n",
    "            if len(r.genes) == 1:\n",
    "                model_rxns_atg.append(r)\n",
    "                model_atg_final.append(atg)\n",
    "            else:\n",
    "                prot_names = []\n",
    "                for g in r.genes:\n",
    "                    prot_names.append(g.id)\n",
    "                if set(prot_names).issubset(set(all_atg)):\n",
    "                    model_rxns_atgs.append(r)\n",
    "                    model_atgs_final = model_atgs_final + prot_names\n",
    "# Feasible constraint summary \n",
    "print(\"Total measured proteins: {}\".format(len(all_atg)))  \n",
    "print(\"Total measured proteins in model: {}\".format(len(model_atg))) \n",
    "print(\"Total model reactions: {}\".format(len(model.reactions)))\n",
    "print(\" \")\n",
    "# Reactions with simple constraints \n",
    "print(\"Total model reactions with single constraints: {}\".format(len(set(model_rxns_atg))))\n",
    "print(\"Model proteins for single constraints: {}\".format(len(set(model_atg_final))))\n",
    "print(\" \")\n",
    "# Reactions with multiple possible constraints \n",
    "print(\"Total model reactions with multiple constraints: {}\".format(len(set(model_rxns_atgs))))\n",
    "print(\"Total model proteins for possible constraints: {}\".format(len(set(model_atgs_final))))\n",
    "# Total constained reactions\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs) #these are the reactions constrained in the models \n",
    "constr_ids = []\n",
    "for r in constr_reacs:\n",
    "    constr_ids.append(r.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting up the Model Constraints\n",
    "# for different genotypes and temperatures \n",
    "\n",
    "def all_constr(C,M,F,S,Cond):\n",
    "    # Metabolite Constraints \n",
    "    print(\"\\n--------- {} -----------\\n\".format(Cond))\n",
    "    f1=0.99*200\n",
    "    f2=1.01*200\n",
    "    rMal = model.reactions.get_by_id('Mal_Store')\n",
    "    rFum = model.reactions.get_by_id('Fum_Store')\n",
    "    rStarch = model.reactions.get_by_id('Starch_Store')\n",
    "    rCO2 = model.reactions.get_by_id('Im_CO2')\n",
    "    FumHA_c = model.reactions.get_by_id('FumHA_c')\n",
    "    rResp = model.reactions.get_by_id('Tr_Pyr2')\n",
    "    rCO2.upper_bound = C+1.0\n",
    "    rCO2.lower_bound = C-1.0\n",
    "    rResp.lower_bound = rCO2.lower_bound*0.15/3\n",
    "    rResp.upper_bound = 1000.0\n",
    "    rMal.upper_bound = (M+0.001)*f2\n",
    "    rMal.lower_bound = (M-0.001)*f1\n",
    "    rFum.upper_bound = (F+0.001)*f2\n",
    "    rFum.lower_bound = (F-0.001)*f1\n",
    "    rStarch.upper_bound = (S-0.001)*f2\n",
    "    rStarch.lower_bound = (S-0.001)*f1\n",
    "    # Proteomic Constraints \n",
    "    name1 = \"{} 1\".format(Cond)\n",
    "    name2 = \"{} 2\".format(Cond)\n",
    "    name3 = \"{} 3\".format(Cond)\n",
    "    name4 = \"{} 4\".format(Cond)\n",
    "    constr(constr_reacs,df[name1],df[name2],df[name3],df[name4])\n",
    "    # Mutant Knockout\n",
    "    if \"fum2\" in Cond:\n",
    "        r = model.reactions.get_by_id('FumHA_c')\n",
    "        r.upper_bound = 0.0 # Mutant knockout \n",
    "        r.lower_bound = 0.0 \n",
    "    # Loopless Constraints \n",
    "    print(\"For Loopless...\")\n",
    "    setobjective(\"Mal_Store\",obj=\"Minimum\")\n",
    "    loopless = loopless_solution(model)\n",
    "    print(loopless[\"Tr_TPT1\"])\n",
    "    print(loopless[\"Tr_TPT2\"])\n",
    "    print(\" \")\n",
    "    for r in model.reactions:\n",
    "        if (loopless[r.id] == 0.0):\n",
    "            r.lower_bound = 0.0\n",
    "            r.upper_bound = 0.0\n",
    "        else:\n",
    "            pass \n",
    "    FVA = flux_variability_analysis(model,['Fd_DASH_NADPR_h'])\n",
    "    NADPH = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "    NADPH.lower_bound = float(FVA['minimum'])-1.0\n",
    "    NADPH.upper_bound = float(FVA['minimum'])+1.0\n",
    "    print(NADPH.bounds)\n",
    "#     FVA = flux_variability_analysis(model,['ATPase_h'])\n",
    "#     ATPase = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "#     ATPase.lower_bound = float(FVA['minimum'])-0.5\n",
    "#     ATPase.upper_bound = float(FVA['minimum'])+0.5\n",
    "#     print(ATPase.bounds)\n",
    "    modelname = \"Arnold2014_{}_LooplessNMin.xml\".format(Cond)\n",
    "    cobra.io.write_sbml_model(model,modelname,use_fbc_package=False)\n",
    "    print(\"MODEL SAVED. Go do flux sampling in MATLAB. \\n\")\n",
    "    # Resetting the mutant knockout \n",
    "    if \"fum2\" in Cond:\n",
    "        r = model.reactions.get_by_id('FumHA_c')\n",
    "        r.upper_bound = 1000.0 \n",
    "        r.lower_bound = 0.0\n",
    "        \n",
    "def AN_constr(C,M,F,S,Cond):\n",
    "    # Metabolite Constraints \n",
    "    print(\"\\n--------- {} -----------\\n\".format(Cond))\n",
    "    f1=0.9*200\n",
    "    f2=1.1*200\n",
    "    rMal = model.reactions.get_by_id('Mal_Store')\n",
    "    rFum = model.reactions.get_by_id('Fum_Store')\n",
    "    rStarch = model.reactions.get_by_id('Starch_Store')\n",
    "    rCO2 = model.reactions.get_by_id('Im_CO2')\n",
    "    FumHA_c = model.reactions.get_by_id('FumHA_c')\n",
    "    rResp = model.reactions.get_by_id('Tr_Pyr2')\n",
    "    ATPase = model.reactions.get_by_id('ATPase_h')\n",
    "    NADPH = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "    rCO2.upper_bound = C*0.01\n",
    "    rCO2.lower_bound = C*0.99\n",
    "    rMal.upper_bound = M*f2\n",
    "    rMal.lower_bound = M*f1\n",
    "    rFum.upper_bound = F*f2\n",
    "    rFum.lower_bound = F*f1\n",
    "    rStarch.upper_bound = S*f2\n",
    "    rStarch.lower_bound = S*f1\n",
    "    # Proteomic Constraints \n",
    "    name1 = \"{} 1\".format(Cond)\n",
    "    name2 = \"{} 2\".format(Cond)\n",
    "    name3 = \"{} 3\".format(Cond)\n",
    "    name4 = \"{} 4\".format(Cond)\n",
    "    constr(constr_reacs,df[name1],df[name2],df[name3],df[name4])\n",
    "    # Mutant Knockout\n",
    "    if \"fum2\" in Cond:\n",
    "        r = model.reactions.get_by_id('FumHA_c')\n",
    "        r.upper_bound = 0.0 # Mutant knockout \n",
    "        r.lower_bound = 0.0 \n",
    "#     # Setting ATP/NADPH ratio\n",
    "#     ATPNADPH(100,145)\n",
    "#     print(\"ATP/NADPH ratio is {}\".format(ATPase.lower_bound/NADPH.upper_bound))\n",
    "    # Loopless Constraints \n",
    "    print(\"For Loopless...\")\n",
    "    setobjective(\"Im_CO2\",obj=\"Maximum\")\n",
    "    loopless = loopless_solution(model)\n",
    "    print(loopless[\"Tr_TPT1\"])\n",
    "    print(loopless[\"Tr_TPT2\"])\n",
    "    print(\" \")\n",
    "    for r in model.reactions:\n",
    "        if (loopless[r.id] == 0.0):\n",
    "            r.lower_bound = 0.0\n",
    "            r.upper_bound = 0.0\n",
    "        else:\n",
    "            pass \n",
    "    # Saving the Model\n",
    "    print(flux_variability_analysis(model,['Im_CO2']))\n",
    "    modelname = \"Arnold2014_{}_Loopless.xml\".format(Cond)\n",
    "    cobra.io.write_sbml_model(model,modelname,use_fbc_package=False)\n",
    "    print(\"MODEL SAVED. Go do flux sampling in MATLAB. \\n\")\n",
    "    # Resetting the mutant knockout \n",
    "    if \"fum2\" in Cond:\n",
    "        r = model.reactions.get_by_id('FumHA_c')\n",
    "        r.upper_bound = 1000.0 \n",
    "        r.lower_bound = 0.0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Genotype and Condition\n",
    "# all_constr(100,0.003,0.006,0.028,\"WT Cntl\")\n",
    "# all_constr(100,0.012,0.0,0.042,\"fum2 Cntl\")\n",
    "# all_constr(100,0.012,0.023,0.051,\"WT cold\")\n",
    "# all_constr(94,0.025,0.0,0.063,\"fum2 Cold\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
