{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(ATCG00020 and ATCG00680 and ATCG00280 and ATCG00270 and ATCG00580 and ATCG00570 and ATCG00710 and ATCG00080 and ATCG00550 and ATCG00070 and ATCG00560 and ATCG00220 and ATCG00700 and (AT5G66570 or AT3G50820) and AT1G06680 and (AT4G21280 or AT4G05180) and AT1G79040 and AT1G44575 and ATCG00690 and AT3G21055 and AT2G30570 and AT2G06520 and AT1G67740 and ATCG00300)' for <Reaction PSII_h at 0x7f74543a7ed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(ATCG00540 and ATCG00720 and AT4G03280 and ATCG00730 and ATCG00600 and ATCG00590 and AT2G26500 and ATCG00210)' for <Reaction Cytb6f_h at 0x7f74543a7e90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(ATCG00490) and 8*(AT5G38430 or AT5G38420 or AT1G67090 or AT5G38410)' for <Reaction RBC_h at 0x7f74543ad310>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G26650 or AT1G12900) or (2*(AT3G26650 or AT1G12900) and 2*(AT1G42970))' for <Reaction GAPDH1_h at 0x7f74543ade90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55440 or AT2G21170)' for <Reaction TPI_h at 0x7f74543ad150>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G01140 or AT4G26530 or AT3G52930 or AT4G38970 or AT5G03690 or AT2G21330)' for <Reaction FBPA_h at 0x7f74543ad190>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G54050)' for <Reaction FBPase_h at 0x7f74543ad250>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G60750 or AT2G45290)' for <Reaction FTK_h at 0x7f74543ad590>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G01140 or AT4G26530 or AT3G52930 or AT4G38970 or AT5G03690 or AT2G21330)' for <Reaction SBPA_h at 0x7f74543ad690>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55800)' for <Reaction SBPase_h at 0x7f74543ad290>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G60750 or AT2G45290)' for <Reaction STK_h at 0x7f74543ad6d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G61410)' for <Reaction Ru5PE_h at 0x7f74543ad710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G04790)' for <Reaction R5PI_h at 0x7f74543ad790>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G32060)' for <Reaction Ru5PK_h at 0x7f74543bc590>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G24620)' for <Reaction PGI_h at 0x7f74543bca50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G19220 or AT1G27680 or AT4G39210 or AT2G21590) and 2*(AT5G48300 or AT1G05610)' for <Reaction AGPase_h at 0x7f74543bce10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55440 or AT2G21170)' for <Reaction TPI_c at 0x7f7454356e90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G26520 or AT4G26530 or AT3G52930 or AT2G36460 or AT5G03690)' for <Reaction FBPA_c at 0x7f7454356f50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G43670)' for <Reaction FBPase_c at 0x7f7454356dd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G07110)' for <Reaction F6PK2_c at 0x7f74543693d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G07110)' for <Reaction F26BPPh_c at 0x7f7454369190>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G24620 or AT5G42740)' for <Reaction PGI_c at 0x7f7454369350>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G20280) or 4*(AT5G20280)' for <Reaction S6PS_c at 0x7f7454369610>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G35840 or AT3G52340)' for <Reaction S6PPh_c at 0x7f74543691d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT5G20830 or AT5G49190 or AT3G43190)' for <Reaction SucS_c at 0x7f74543699d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G53310 or AT2G42600 or AT3G14940 or AT1G68750)' for <Reaction PEPC2_c at 0x7f7454369dd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G29560 or AT2G36530)' for <Reaction Enol_c at 0x7f7454369e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G04120 or AT1G13440)' for <Reaction GAPDH2_c at 0x7f745437e050>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G79530 or AT1G16300)' for <Reaction GAPDH2_h at 0x7f7454312190>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G74030 or AT2G36530)' for <Reaction Enol_h at 0x7f7454312750>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G04050 or AT5G08570 or AT3G25960 or AT2G36580 or AT3G55650 or AT3G55810 or AT5G63680 or AT3G52990)' for <Reaction PyrK_c at 0x7f7454312710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G22960) and 4*(AT5G52920 or AT1G32440)' for <Reaction PyrK_h at 0x7f7454312950>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G59900) and 2*(AT5G50850)' for <Reaction PyrDH1_m at 0x7f7454312b50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G59900) and 2*(AT5G50850)' for <Reaction PyrDH2_m at 0x7f7454312ad0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT3G52200 or AT3G13930 or AT1G54220)' for <Reaction DHLAcT_m at 0x7f7454312e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G48030 or AT3G17240)' for <Reaction DHLDH2_m at 0x7f7454312f50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G65750 or AT3G55410)' for <Reaction KGDH_m at 0x7f7454312890>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT5G55070 or AT4G26910)' for <Reaction DHLST_m at 0x7f7454312a10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(ATMG01275 or ATMG01120 or ATMG00516) and (ATMG00285 or ATMG01320) and ATMG00990 and (ATMG00580 or ATMG00650) and (ATMG00513 or ATMG00665 or ATMG00060) and ATMG00270 and ATMG00510 and ATMG00070 and AT5G37510 and AT5G11770 (AT1G79010 or AT1G16700) and AT5G08530 and AT4G02580' for <Reaction cplx1_m at 0x7f7454326590>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(ATMG01190) and 3*(AT5G08670 or AT5G08690 or AT5G08680) and AT2G33040 and (AT5G13450 or AT5G47030) and AT1G51650 and (ATMG00410 or ATMG01170) and 2*(ATMG00640 or ATMG00480) and 12*(ATMG01080 or AT2G07671) and AT3G52300' for <Reaction cplx5_m at 0x7f7454326c10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(ATCG00490) and 8*(AT5G38430 or AT5G38420 or AT1G67090 or AT5G38410)' for <Reaction RBO_h at 0x7f7454326d50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G20620 or AT1G20630 or AT4G35090)' for <Reaction CAT_p at 0x7f7454326e50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G33010 or AT2G26080)' for <Reaction GlyDH_m at 0x7f7454335750>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G48030 or AT3G17240)' for <Reaction DHLDH1_m at 0x7f74543353d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G37930 or AT5G26780)' for <Reaction GlyHMT_m at 0x7f74542c8350>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G13360)' for <Reaction SGAT_p at 0x7f74542c8bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G19900 or AT5G11670 or AT5G25880)' for <Reaction MalDH2NADP_c at 0x7f74542df9d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT5G11670 or AT5G25880 or AT1G79750)' for <Reaction MalDH2NADP_h at 0x7f74542dfb90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G10670 or AT1G60810 or AT1G09430) and 4*(AT3G06650 orAT5G49460)' for <Reaction ATPCitL_c at 0x7f74542f4710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT5G16390 or AT5G15530) and AT5G35360 and 2*(AT2G38040) and 2*(ATCG00500)' for <Reaction ACoAC_h at 0x7f74542f4d90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G35460 or AT2G17420 or AT2G41680)' for <Reaction TRXR_c at 0x7f745428a390>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G12440 or AT5G11160 or AT4G22570 or AT1G27450 or AT1G80050)' for <Reaction ADPRPPT_c at 0x7f745428a710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G77760 or AT1G37130)' for <Reaction NO3R1_c at 0x7f745428ac50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G77760 or AT1G37130)' for <Reaction NO3R2_c at 0x7f745428aa90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G22890 or AT1G19920 or AT4G14680 or AT5G43780)' for <Reaction ATPSL_h at 0x7f745428af90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G54660)' for <Reaction GSSGR_h at 0x7f745428add0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT1G08065 or AT5G04180 or AT4G20990 or AT4G21000 or AT2G28210 or AT3G52720) or 8*(AT1G23730 or AT5G14740)' for <Reaction HCO3DHA_c at 0x7f745428ad90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT4G20990 or AT1G08080 or AT3G52720) or 8*(AT1G23730 or AT1G70410 or AT1G58180 or AT3G01500 or AT5G14740 or AT4G33580)' for <Reaction HCO3DHA_h at 0x7f745428af10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT3G57560)' for <Reaction AGluK at 0x7f745429e090>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G08100) and 2*(AT3G16150)' for <Reaction Asnase_c at 0x7f74542b6250>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT2G17640 or AT3G13110 or AT4G35640 or AT5G56760)' for <Reaction SerAcT_c at 0x7f74542b6490>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT3G13110)' for <Reaction SerAcT_h at 0x7f74542b6810>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT3G13110)' for <Reaction SerAcT_m at 0x7f74542b6910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G14880 or AT3G22460)' for <Reaction CysS_c at 0x7f74542b6a50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G61440 or AT3G03630 or AT4G14880 or AT2G43750)' for <Reaction CysS_h at 0x7f74542b6b90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G61440 or AT3G03630 or AT2G43750 or AT5G28020)' for <Reaction CysS_m at 0x7f74542b6c90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440 or AT3G03910)' for <Reaction GluDH1NAD_m at 0x7f74542b6710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440 or AT3G03910)' for <Reaction GluDH1NADP_m at 0x7f745424a090>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440)' for <Reaction GluDH2NAD_m at 0x7f74542b6b50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G17330 or AT1G65960 or AT2G02000 or AT2G02010 or AT3G17760)' for <Reaction GluDC_c at 0x7f745424a490>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G22200)' for <Reaction GABATA1_m at 0x7f745424a610>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G22200)' for <Reaction GABATA2_m at 0x7f745424a790>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G79440)' for <Reaction SCASeADH_m at 0x7f745424a590>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G37600 or AT3G17820 or AT5G16570 or AT1G66200)' for <Reaction GlnS_c at 0x7f745424a750>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT3G17820 or AT1G48470 or AT5G35630)' for <Reaction GlnS_h at 0x7f745424a990>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G35630)' for <Reaction GlnS_m at 0x7f745424ab10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G20620 or AT1G20630 or AT4G35090)' for <Reaction CAT_h at 0x7f745424a310>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G13360' for <Reaction SGAT_h at 0x7f745424a710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G10050)' for <Reaction ThrDA_h at 0x7f745425f910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS1_h at 0x7f745425f9d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS2_h at 0x7f745425fb90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS3_h at 0x7f745425fdd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G18500) or 4*(AT1G74040)' for <Reaction 2IPMS_h at 0x7f7454271290>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G19710 or AT1G31230 or AT5G13280 or AT3G02020 or AT5G14060)' for <Reaction AspK_h at 0x7f7454271310>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G14810)' for <Reaction Asp_DASH_SeADH_h at 0x7f74542714d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G33680)' for <Reaction DAPAT_h at 0x7f7454271910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G31230 AT4G19710)' for <Reaction HSerDHNADP_h at 0x7f7454271ed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G31230 AT4G19710)' for <Reaction HSerDHNAD_h at 0x7f7454271b50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G57050)' for <Reaction CTHL_h at 0x7f7454209310>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G44160 or AT3G59970)' for <Reaction 5M_DASH_THFor_c at 0x7f7454209110>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G13930 or AT4G13890 or AT1G22020 or AT1G36370 or AT4G37930 or AT5G26780)' for <Reaction GlyHMT_c at 0x7f7454209790>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G29840 or AT1G72810)' for <Reaction ThrS_h at 0x7f745422f150>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G24807) and 2*(AT3G55870 or AT5G05730 or AT2G29690)' for <Reaction ANTS_h at 0x7f745422f110>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G38280)' for <Reaction AMPDA_c at 0x7f7454242250>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G57610)' for <Reaction DCAMPS_h at 0x7f7454242390>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction ADPR_c at 0x7f7454242e90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction CDPR_c at 0x7f7454242e50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction GDPR_c at 0x7f7454242fd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction UDPR_c at 0x7f74541d8610>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT3G46940)' for <Reaction dUTPP_c at 0x7f74541d8750>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction GDPK_c at 0x7f74541d8350>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction GDPK_h at 0x7f74541d8710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction UDPK_c at 0x7f74541d8c90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dADPK_c at 0x7f74541d8dd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dCDPK_c at 0x7f74541d8f10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dGDPK_c at 0x7f74541d8ed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dTDPK_c at 0x7f74541d8bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dUDPK_c at 0x7f74541d84d0>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mal_c --> Fum_c + H2O_c\n",
      "(0.0, 1000.0)\n",
      "Directionality changed. Dyson et al. 2016 \n",
      "\n",
      "Mal_c --> \n",
      "Fum_c --> \n",
      "starch1_h --> \n"
     ]
    }
   ],
   "source": [
    "# Loading Required Packages \n",
    "import cobra\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "# Loading the Model \n",
    "model = cobra.io.read_sbml_model(\"ArabidopsisCoreModel1.xml\")\n",
    "\n",
    "# Changing the directionality of Fum_c to Mal_c to go the other way (day-time model) \n",
    "r = model.reactions.get_by_id('FumHA_c')\n",
    "r.add_metabolites({model.metabolites.get_by_id('Fum_c'): +2})\n",
    "r.add_metabolites({model.metabolites.get_by_id('H2O_c'): +2})\n",
    "r.add_metabolites({model.metabolites.get_by_id('Mal_c'): -2})\n",
    "print(r.reaction)\n",
    "print(r.bounds)\n",
    "print(\"Directionality changed. Dyson et al. 2016 \\n\")\n",
    "\n",
    "# Creating a Malate \"Storage\" Reactions\n",
    "r = cobra.Reaction('Mal_Store')\n",
    "r.name = 'Mal_Store'\n",
    "r.add_metabolites({model.metabolites.get_by_id('Mal_c'): -1})\n",
    "model.add_reaction(r)\n",
    "print(r.reaction)\n",
    "\n",
    "# Creating a Fumarate \"Storage\" Reactions\n",
    "r = cobra.Reaction('Fum_Store')\n",
    "r.name = 'Fum_Store'\n",
    "r.add_metabolites({model.metabolites.get_by_id('Fum_c'): -1})\n",
    "model.add_reaction(r)\n",
    "print(r.reaction)\n",
    "\n",
    "# Creating a Starch \"Storage\" Reactions\n",
    "r = cobra.Reaction('Starch_Store')\n",
    "r.name = 'Starch_Store'\n",
    "r.add_metabolites({model.metabolites.get_by_id('starch1_h'): -1})\n",
    "model.add_reaction(r)\n",
    "print(r.reaction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Printing all import and export reactions in the model\n",
    "def input_output_reacs(model):\n",
    "    \"Prints all input and output reactions for the given model\"\n",
    "    reacs = []\n",
    "    for r in model.reactions:\n",
    "        coeffs = []\n",
    "        for i in r.metabolites:\n",
    "            c = r.get_coefficient(i.id)\n",
    "            coeffs.append(c)\n",
    "        if all(i > 0.0 for i in coeffs) or all(i < 0.0 for i in coeffs) :\n",
    "            reacs.append(r)\n",
    "    return(reacs)\n",
    "\n",
    "exp_reacs = input_output_reacs(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IPODC_h\n",
      "H_h + IPO_h --> 4MOP_h + CO2_h\n",
      "GluSeADA_c\n",
      "Glu_DASH_SeA_c --> H2O_c + H_c + P5C_c\n",
      "GluSeADA_h\n",
      "Glu_DASH_SeA_h --> H2O_h + H_h + P5C_h\n",
      "GluSeADA_m\n",
      "Glu_DASH_SeA_m --> H2O_m + H_m + P5C_m\n",
      "Im_Pi\n",
      "3.0 ATP_c + 3.0 H2O_c --> 3.0 ADP_c + 2.0 H_h + 3.0 Pi_c + Pi_h\n",
      "Im_NO3\n",
      "2.0 ATP_c + 2.0 H2O_c --> 2.0 ADP_c + 2.0 H_c + NO3_c + 2.0 Pi_c\n",
      "Im_SO4\n",
      "3.0 ATP_c + 3.0 H2O_c --> 3.0 ADP_c + 3.0 H_c + 3.0 Pi_c + SO4_c\n",
      "Ex_MACP\n",
      "M_DASH_ACP_h --> ACP_h + H_h\n",
      "NGAM_c\n",
      "ATP_c + H2O_c --> ADP_c + H_c + Pi_c\n",
      "NGAM_h\n",
      "ATP_h + H2O_h --> ADP_h + H_h + Pi_h\n",
      "NGAM_m\n",
      "ATP_m + H2O_m --> ADP_m + H_m + Pi_m\n",
      "Bio_CLim\n",
      "0.58559 ATP_c + 207.46087 Ala_c + 64.19766 Arg_c + 61.82037 Asn_c + 90.68407 Asp_c + 0.49638 CTP_c + 27.91206 Cys_c + 14.77415 Frc_c + 77.58391 Fum_m + 0.15305 GABA_c + 0.40478 GTP_c + 33.84942 Glc_c + 45.7267 Gln_c + 103.37619 Glu_c + 245.76971 Gly_c + 29.81972 His_c + 80.57171 Ile_c + 142.42163 Leu_c + 92.89756 Lys_c + 779.97036 M_DASH_ACP_h + 38.11199 Mal_m + 2.27742 Mas_c + 30.37981 Met_c + 14.60727 Orn_h + 59.65081 Phe_c + 96.88322 Pro_c + 0.31189 SA_h + 1.30199 SCA_m + 132.78221 Ser_c + 45.39823 Suc_c + 125.42878 Thr_c + 1.15818 Tre_c + 15.08586 Trp_c + 42.33459 Tyr_c + 0.61307 UTP_c + 131.21257 Val_c + 363.88306 cellulose2_c + 0.68091 dATP_c + 0.40366 dCTP_c + 0.37141 dGTP_c + 0.69281 dTTP_c + 227.34668 starch2_h + 5.53164 urea_m --> 779.97036 ACP_h + 2673.89 H_h\n",
      "Bio_NLim\n",
      "0.62086 ATP_c + 351.01628 Ala_c + 142.53474 Arg_c + 110.85014 Asn_c + 154.48941 Asp_c + 0.52628 CTP_c + 47.36239 Cys_c + 10.66909 Frc_c + 11.5874 Fum_m + 0.58323 GABA_c + 0.42916 GTP_c + 23.53917 Glc_c + 84.12036 Gln_c + 176.89469 Glu_c + 415.15779 Gly_c + 50.59939 His_c + 136.84188 Ile_c + 241.74348 Leu_c + 159.10723 Lys_c + 826.95653 M_DASH_ACP_h + 10.89856 Mal_m + 5.37965 Mas_c + 51.81818 Met_c + 9.64682 Orn_h + 101.65645 Phe_c + 164.00331 Pro_c + 0.19059 SA_h + 1.62992 SCA_m + 224.02319 Ser_c + 87.01511 Suc_c + 204.61559 Thr_c + 2.82436 Tre_c + 26.64807 Trp_c + 71.73651 Tyr_c + 0.65 UTP_c + 222.6177 Val_c + 363.88306 cellulose2_c + 0.72193 dATP_c + 0.42798 dCTP_c + 0.39378 dGTP_c + 0.73455 dTTP_c + 335.54269 starch2_h + 10.36535 urea_m --> 826.95653 ACP_h + 5581.92 H_h\n",
      "Bio_opt\n",
      "0.58559 ATP_c + 230.67099 Ala_c + 74.68248 Arg_c + 70.5431 Asn_c + 101.03805 Asp_c + 0.49638 CTP_c + 31.11756 Cys_c + 12.0349 Frc_c + 67.20727 Fum_m + 0.09401 GABA_c + 0.40478 GTP_c + 18.47868 Glc_c + 51.2643 Gln_c + 116.39401 Glu_c + 273.48273 Gly_c + 33.2443 His_c + 89.82686 Ile_c + 158.77115 Leu_c + 103.59352 Lys_c + 779.97036 M_DASH_ACP_h + 50.99641 Mal_m + 2.60958 Mas_c + 33.61497 Met_c + 1.58654 Orn_h + 66.66173 Phe_c + 108.03688 Pro_c + 0.37924 SA_h + 0.98709 SCA_m + 147.49851 Ser_c + 30.66837 Suc_c + 136.6831 Thr_c + 0.57736 Tre_c + 16.87103 Trp_c + 47.02921 Tyr_c + 0.61307 UTP_c + 146.26883 Val_c + 363.88306 cellulose2_c + 0.68091 dATP_c + 0.40366 dCTP_c + 0.37141 dGTP_c + 0.69281 dTTP_c + 294.86155 starch2_h + 2.84043 urea_m --> 779.97036 ACP_h + 3125.54 H_h\n",
      "136\n"
     ]
    }
   ],
   "source": [
    "# Appending transport reactions which are not to be deleted\n",
    "r1 = model.reactions.get_by_id(\"Tr_TPT1\")\n",
    "r2 = model.reactions.get_by_id(\"Tr_TPT2\")\n",
    "\n",
    "r = r1\n",
    "r.upper_bound = 0.0\n",
    "r.lower_bound = 0.0\n",
    "\n",
    "\n",
    "# Setting reactions without AT codes to zero\n",
    "noAT = []\n",
    "for r in model.reactions:\n",
    "    g = r.genes\n",
    "    for gen in g:\n",
    "        if gen.id == '-':\n",
    "            if r not in exp_reacs:\n",
    "#                 if r.id[0] is not \"T\":\n",
    "#                     print(r.id)\n",
    "#                     print(r.reaction)\n",
    "#                     r.upper_bound = 0.0\n",
    "#                     r.lower_bound = 0.0\n",
    "                noAT.append(r)\n",
    "print(len(noAT))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of all possible proteins in the model\n",
    "gene_ids = []\n",
    "for g in model.genes:\n",
    "    gene_ids.append(g.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reading in the protein data\n",
    "df = pd.read_excel(io=\"ProteinConc.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of all measured proteins \n",
    "all_atg = []\n",
    "x=df[\"Accession\"]\n",
    "for atg in x.values:\n",
    "    all_atg.append(atg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measured proteins: 2427\n",
      "Total measured proteins in model: 245\n",
      "Total model reactions: 552\n",
      " \n",
      "Total model reactions with single constraints: 62\n",
      "Model proteins for single constraints: 49\n",
      " \n",
      "Total model reactions with multiple constraints: 39\n",
      "Total model proteins for possible constraints: 53\n"
     ]
    }
   ],
   "source": [
    "# Comparing the measured protein concentrations to those which are in the model \n",
    "\n",
    "model_atg = [] # Total measured proteins in model\n",
    "model_atg_final = [] # Proteins with direct reaction constraint (i.e. single responsible protein)\n",
    "model_rxns_atg = [] # Model reactions with direct constraints (i.e. single responsible protein)\n",
    "model_rxns_atgs = [] # Model reactions with multiple possible constraints \n",
    "model_atgs_final = [] # Proteins with multiple possible constraints \n",
    "\n",
    "for atg in x.values: # Check if protein is in the model \n",
    "    if atg in gene_ids:\n",
    "        model_atg.append(atg)\n",
    "        # Only parameterize that reaction if that enzyme is the only one responsible for that reaction\n",
    "        for r in model.genes.get_by_id(atg).reactions:\n",
    "            if len(r.genes) == 1:\n",
    "                model_rxns_atg.append(r)\n",
    "                model_atg_final.append(atg)\n",
    "            else:\n",
    "                prot_names = []\n",
    "                for g in r.genes:\n",
    "                    prot_names.append(g.id)\n",
    "                if set(prot_names).issubset(set(all_atg)):\n",
    "                    model_rxns_atgs.append(r)\n",
    "                    model_atgs_final = model_atgs_final + prot_names\n",
    "\n",
    "# Feasible constraint summary \n",
    "# General \n",
    "print(\"Total measured proteins: {}\".format(len(all_atg)))  \n",
    "print(\"Total measured proteins in model: {}\".format(len(model_atg))) \n",
    "print(\"Total model reactions: {}\".format(len(model.reactions)))\n",
    "print(\" \")\n",
    "# Reactions with simple constraints \n",
    "print(\"Total model reactions with single constraints: {}\".format(len(set(model_rxns_atg))))\n",
    "print(\"Model proteins for single constraints: {}\".format(len(set(model_atg_final))))\n",
    "print(\" \")\n",
    "# Reactions with multiple possible constraints \n",
    "print(\"Total model reactions with multiple constraints: {}\".format(len(set(model_rxns_atgs))))\n",
    "print(\"Total model proteins for possible constraints: {}\".format(len(set(model_atgs_final))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "constr_ids = []\n",
    "for r in constr_reacs:\n",
    "    constr_ids.append(r.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function for constraining the model reactions for a specific genotype and temperature\n",
    "def constr(reacs,Dat1,Dat2,Dat3,Dat4):\n",
    "    for r in reacs:\n",
    "        upper_vals = []\n",
    "        lower_vals = []\n",
    "        for g in r.genes:\n",
    "            # Get index of protein in data file \n",
    "            ind = all_atg.index(g.id)\n",
    "            # Scaling the values for setting fluxes \n",
    "            vals = [i/4000. for i in Dat1[ind],Dat2[ind],Dat3[ind],Dat4[ind]]\n",
    "            val = np.mean(vals)\n",
    "            sterr = np.std(vals)/np.sqrt(len(vals))\n",
    "            upper = val + sterr\n",
    "            lower = val - sterr\n",
    "            upper_vals.append(upper)\n",
    "            lower_vals.append(lower)\n",
    "        # Constraint flux of forward reaction\n",
    "        r.upper_bound = round(sum(upper_vals),3)\n",
    "        # Constraint flux of lower reaction \n",
    "        if r.lower_bound < 0.0:\n",
    "            r.lower_bound = round(-sum(upper_vals),3)\n",
    "        # No lower bound b/c don't know localization or enzymes are used to in other reactions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " --> CO2_c\n",
      "Maximize\n",
      "-1.0*Im_CO2_reverse_a674e + 1.0*Im_CO2\n"
     ]
    }
   ],
   "source": [
    "# RUN THIS NEXT SECTION AS EITHER OR! RESERT KERNEL EACH TIME. \n",
    "\n",
    "rMal = model.reactions.get_by_id('Mal_Store')\n",
    "rFum = model.reactions.get_by_id('Fum_Store')\n",
    "rStarch = model.reactions.get_by_id('Starch_Store')\n",
    "rCO2 = model.reactions.get_by_id('Im_CO2')\n",
    "FumHA_c = model.reactions.get_by_id('FumHA_c')\n",
    "NADPH = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "ATPase = model.reactions.get_by_id('ATPase_h')\n",
    "all_reac = [rMal,rFum,rStarch,rCO2,FumHA_c,NADPH,ATPase]\n",
    "rObj = model.reactions.get_by_id('Im_CO2')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "\n",
    "f1=0.99*200\n",
    "f2=1.01*200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    maximum     minimum\n",
      "Mal_Store          0.808000    0.396000\n",
      "Fum_Store          1.414000    0.990000\n",
      "Starch_Store       5.858000    5.346000\n",
      "Im_CO2           101.000000  101.000000\n",
      "FumHA_c            1.414000    0.000000\n",
      "Fd_DASH_NADPR_h  250.000000  141.873429\n",
      "ATPase_h         107.142857   85.551000\n"
     ]
    }
   ],
   "source": [
    "# Parameterizing the model for Col-0 Control Conditions\n",
    "\n",
    "# Col0 Control\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rMal.upper_bound = 0.004*f2\n",
    "rMal.lower_bound = 0.002*f1\n",
    "rFum.upper_bound = 0.007*f2\n",
    "rFum.lower_bound = 0.005*f1\n",
    "rStarch.upper_bound = 0.029*f2\n",
    "rStarch.lower_bound = 0.027*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "constr(constr_reacs,df[\"WT Cntl 1\"],df[\"WT Cntl 2\"],df[\"WT Cntl 3\"],df[\"WT Cntl 4\"])\n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cntl_All_removeAT.xml\",use_fbc_package=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Fum2 Control Conditions\n",
    "\n",
    "# Fum2 Control\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rMal.upper_bound = 0.013*f2\n",
    "rMal.lower_bound = 0.011*f1\n",
    "rFum.upper_bound = 0.0*f2\n",
    "rFum.lower_bound = 0.0*f1\n",
    "rStarch.upper_bound = 0.043*f2\n",
    "rStarch.lower_bound = 0.041*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "r = model.reactions.get_by_id('FumHA_c')\n",
    "r.upper_bound = 0.0 # Mutant knockout \n",
    "r.lower_bound = 0.0 \n",
    "constr(constr_reacs,df[\"fum2 Cntl 1\"],df[\"fum2 Cntl 2\"],df[\"fum2 Cntl 3\"],df[\"fum2 Cntl 4\"]) #Proteomics \n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_M_Cntl_All_removeAT.xml\",use_fbc_package=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Col-0 Cold Conditions\n",
    "\n",
    "# Col0 Cold\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rMal.upper_bound = 0.013*f2\n",
    "rMal.lower_bound = 0.011*f1\n",
    "rFum.upper_bound = 0.024*f2\n",
    "rFum.lower_bound = 0.022*f1\n",
    "rStarch.upper_bound = 0.052*f2\n",
    "rStarch.lower_bound = 0.05*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "constr(constr_reacs,df[\"WT cold 1\"],df[\"WT cold 2\"],df[\"WT cold 3\"],df[\"WT cold 4\"])\n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cold_All.xml\",use_fbc_package=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Fum2 Cold Conditions\n",
    "\n",
    "# Fum2 Cold\n",
    "rCO2.upper_bound = 95.0\n",
    "rCO2.lower_bound = 93.0\n",
    "rMal.upper_bound = 0.026*f2\n",
    "rMal.lower_bound = 0.024*f1\n",
    "rFum.upper_bound = 0.0*f2\n",
    "rFum.lower_bound = 0.0*f1\n",
    "rStarch.upper_bound = 0.064*f2\n",
    "rStarch.lower_bound = 0.062*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "r = model.reactions.get_by_id('FumHA_c')\n",
    "r.upper_bound = 0.0 # Mutant knockout \n",
    "r.lower_bound = 0.0 \n",
    "constr(constr_reacs,df[\"fum2 Cold 1\"],df[\"fum2 Cold 2\"],df[\"fum2 Cold 3\"],df[\"fum2 Cold 4\"]) #Proteomics \n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_M_Cold_All.xml\",use_fbc_package=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathwayA = [\"Tr_TPT1\"]\n",
    "pathwayB = [\"PGAK_h\",\"GAPDH2_h\",\"Tr_TPT2\",\"GAPDH3_c\"]\n",
    "both = [\"RBC_h\",\"PGAM_c\",\"Enol_c\",\"PEPC2_c\",\"MalDH1_c\",\"FumHA_c\"]\n",
    "allnames = both+pathwayB+pathwayA\n",
    "for rname in allnames:\n",
    "    r=model.reactions.get_by_id(rname)\n",
    "    print(r.id)\n",
    "    print(r.reaction)\n",
    "    all_reac.append(rname)\n",
    "\n",
    "\n",
    "rObj = model.reactions.get_by_id('Im_CO2')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "rObj = model.reactions.get_by_id('ATPase_h')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Minimum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "rObj = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Minimum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHRR flux sampling (MATLAB SCRIPT)\n",
    "# Plotting the overlapping sampling distribution (NEW PYTHON SCRIPT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting model reactions as sets \n",
    "req_reacs = []\n",
    "Nreq_reacs = []\n",
    "\n",
    "# Exluding reactions which cannot be deleted in combination with others\n",
    "# Round 1\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2h\"))\n",
    "# Round 2 \n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_CO2h\"))\n",
    "# Round 3\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_H2Oh\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Im_NO3\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2m\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_H2Oh\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2m\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_NO2\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_NTT\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_AAC\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT1\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT2\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT3\"))\n",
    "\n",
    "noAT = []\n",
    "for r in model.reactions:\n",
    "    g = r.genes\n",
    "    for gen in g:\n",
    "        if gen.id == '-':\n",
    "            if r not in exp_reacs:\n",
    "                noAT.append(r)\n",
    "                r.upper_bound = 0.0\n",
    "                r.lower_bound = 0.0\n",
    "                \n",
    "print(len(noAT))\n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "# # Trying to delete as high a set of them as possible \n",
    "# for s in set(itertools.combinations(noAT, 4)): #set(itertools.combinations(noAT, 1))\n",
    "#     print(\"Check\")\n",
    "#     b = []\n",
    "#     for i in range(len(s)):\n",
    "#         b.append(s[i].bounds)\n",
    "#         s[i].upper_bound = 0.0\n",
    "#         s[i].lower_bound = 0.0\n",
    "#     interest_reacs = [\"Im_CO2\",\"Mal_Store\",\"Fum_Store\",\"Starch_Store\",\"Ex_Suc\"]\n",
    "#     test = []\n",
    "#     for r in interest_reacs:\n",
    "#         model.objective = r\n",
    "#         try:\n",
    "#             bounds = flux_variability_analysis(model, r)\n",
    "#             if bounds['minimum'][0] > 0.0:\n",
    "#                 test.append(0) # Carbon storage ok\n",
    "#             else:\n",
    "#                 test.append(1)\n",
    "#         except Exception:\n",
    "#             test.append(1)\n",
    "#     if sum(test) > 0: # Can't store carbon \n",
    "#         req_reacs.append(s)\n",
    "#     else:\n",
    "#         Nreq_reacs.append(s)\n",
    "#     for i in range(len(s)):\n",
    "#         # Resetting the reactions bounds\n",
    "#         s[i].upper_bound = b[i][1]\n",
    "#         s[i].lower_bound = b[i][0]\n",
    "# print(\"Finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting model reactions as sets \n",
    "req_reacs = []\n",
    "Nreq_reacs = []\n",
    "\n",
    "# Trying to delete as high a set of them as possible \n",
    "for s in set(itertools.combinations(model.reactions, 1)): #set(itertools.combinations(noAT, 1))\n",
    "    b = []\n",
    "    for i in range(len(s)):\n",
    "        b.append(s[i].bounds)\n",
    "        s[i].upper_bound = 0.0\n",
    "        s[i].lower_bound = 0.0\n",
    "    interest_reacs = [\"Im_CO2\",\"Mal_Store\",\"Starch_Store\",\"Ex_Suc\"]\n",
    "    test = []\n",
    "    for r in interest_reacs:\n",
    "        model.objective = r\n",
    "        try:\n",
    "            bounds = flux_variability_analysis(model, r)\n",
    "            if bounds['minimum'][0] > 0.0:\n",
    "                test.append(0) # Carbon storage ok\n",
    "            else:\n",
    "                test.append(1)\n",
    "        except Exception:\n",
    "            test.append(1)\n",
    "    if sum(test) > 0: # Can't store carbon \n",
    "        req_reacs.append(s)\n",
    "    else:\n",
    "        Nreq_reacs.append(s)\n",
    "    for i in range(len(s)):\n",
    "        # Resetting the reactions bounds\n",
    "        s[i].upper_bound = b[i][1]\n",
    "        s[i].lower_bound = b[i][0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
