{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading Required Packages \n",
    "import cobra\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "from cobra.flux_analysis.loopless import add_loopless, loopless_solution\n",
    "from cobra.flux_analysis import pfba\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the protein data \n",
    "df = pd.read_excel(io=\"ProteinConc.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(ATCG00020 and ATCG00680 and ATCG00280 and ATCG00270 and ATCG00580 and ATCG00570 and ATCG00710 and ATCG00080 and ATCG00550 and ATCG00070 and ATCG00560 and ATCG00220 and ATCG00700 and (AT5G66570 or AT3G50820) and AT1G06680 and (AT4G21280 or AT4G05180) and AT1G79040 and AT1G44575 and ATCG00690 and AT3G21055 and AT2G30570 and AT2G06520 and AT1G67740 and ATCG00300)' for <Reaction PSII_h at 0x7fa538511950>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(ATCG00540 and ATCG00720 and AT4G03280 and ATCG00730 and ATCG00600 and ATCG00590 and AT2G26500 and ATCG00210)' for <Reaction Cytb6f_h at 0x7fa537f55e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(ATCG00490) and 8*(AT5G38430 or AT5G38420 or AT1G67090 or AT5G38410)' for <Reaction RBC_h at 0x7fa5385119d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G26650 or AT1G12900) or (2*(AT3G26650 or AT1G12900) and 2*(AT1G42970))' for <Reaction GAPDH1_h at 0x7fa538511bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55440 or AT2G21170)' for <Reaction TPI_h at 0x7fa538511c10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G01140 or AT4G26530 or AT3G52930 or AT4G38970 or AT5G03690 or AT2G21330)' for <Reaction FBPA_h at 0x7fa538511d10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G54050)' for <Reaction FBPase_h at 0x7fa538511cd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G60750 or AT2G45290)' for <Reaction FTK_h at 0x7fa538377f90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G01140 or AT4G26530 or AT3G52930 or AT4G38970 or AT5G03690 or AT2G21330)' for <Reaction SBPA_h at 0x7fa5383771d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55800)' for <Reaction SBPase_h at 0x7fa538377110>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G60750 or AT2G45290)' for <Reaction STK_h at 0x7fa538377a90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G61410)' for <Reaction Ru5PE_h at 0x7fa538377f50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G04790)' for <Reaction R5PI_h at 0x7fa538377190>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G32060)' for <Reaction Ru5PK_h at 0x7fa53835e050>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G24620)' for <Reaction PGI_h at 0x7fa53835e4d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G19220 or AT1G27680 or AT4G39210 or AT2G21590) and 2*(AT5G48300 or AT1G05610)' for <Reaction AGPase_h at 0x7fa53835e8d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G55440 or AT2G21170)' for <Reaction TPI_c at 0x7fa538388910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G26520 or AT4G26530 or AT3G52930 or AT2G36460 or AT5G03690)' for <Reaction FBPA_c at 0x7fa5383889d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G43670)' for <Reaction FBPase_c at 0x7fa538388cd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G07110)' for <Reaction F6PK2_c at 0x7fa538388e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G07110)' for <Reaction F26BPPh_c at 0x7fa538388bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G24620 or AT5G42740)' for <Reaction PGI_c at 0x7fa538388c50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G20280) or 4*(AT5G20280)' for <Reaction S6PS_c at 0x7fa538388b10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G35840 or AT3G52340)' for <Reaction S6PPh_c at 0x7fa538388b90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT5G20830 or AT5G49190 or AT3G43190)' for <Reaction SucS_c at 0x7fa538388b50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G53310 or AT2G42600 or AT3G14940 or AT1G68750)' for <Reaction PEPC2_c at 0x7fa5383a38d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G29560 or AT2G36530)' for <Reaction Enol_c at 0x7fa5383a3910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G04120 or AT1G13440)' for <Reaction GAPDH2_c at 0x7fa5383b6950>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G79530 or AT1G16300)' for <Reaction GAPDH2_h at 0x7fa5383b6c10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G74030 or AT2G36530)' for <Reaction Enol_h at 0x7fa5383b6cd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G04050 or AT5G08570 or AT3G25960 or AT2G36580 or AT3G55650 or AT3G55810 or AT5G63680 or AT3G52990)' for <Reaction PyrK_c at 0x7fa5383b6e50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G22960) and 4*(AT5G52920 or AT1G32440)' for <Reaction PyrK_h at 0x7fa5383b6e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G59900) and 2*(AT5G50850)' for <Reaction PyrDH1_m at 0x7fa5383cc610>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G59900) and 2*(AT5G50850)' for <Reaction PyrDH2_m at 0x7fa5383cc590>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT3G52200 or AT3G13930 or AT1G54220)' for <Reaction DHLAcT_m at 0x7fa5383cc8d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G48030 or AT3G17240)' for <Reaction DHLDH2_m at 0x7fa5383cc9d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G65750 or AT3G55410)' for <Reaction KGDH_m at 0x7fa5383cc410>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT5G55070 or AT4G26910)' for <Reaction DHLST_m at 0x7fa5383cc310>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(ATMG01275 or ATMG01120 or ATMG00516) and (ATMG00285 or ATMG01320) and ATMG00990 and (ATMG00580 or ATMG00650) and (ATMG00513 or ATMG00665 or ATMG00060) and ATMG00270 and ATMG00510 and ATMG00070 and AT5G37510 and AT5G11770 (AT1G79010 or AT1G16700) and AT5G08530 and AT4G02580' for <Reaction cplx1_m at 0x7fa537cdd050>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(ATMG01190) and 3*(AT5G08670 or AT5G08690 or AT5G08680) and AT2G33040 and (AT5G13450 or AT5G47030) and AT1G51650 and (ATMG00410 or ATMG01170) and 2*(ATMG00640 or ATMG00480) and 12*(ATMG01080 or AT2G07671) and AT3G52300' for <Reaction cplx5_m at 0x7fa537cdd690>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(ATCG00490) and 8*(AT5G38430 or AT5G38420 or AT1G67090 or AT5G38410)' for <Reaction RBO_h at 0x7fa537cdd7d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G20620 or AT1G20630 or AT4G35090)' for <Reaction CAT_p at 0x7fa537cdd8d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G33010 or AT2G26080)' for <Reaction GlyDH_m at 0x7fa537cddc90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G48030 or AT3G17240)' for <Reaction DHLDH1_m at 0x7fa537cdde50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G37930 or AT5G26780)' for <Reaction GlyHMT_m at 0x7fa537cefdd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G13360)' for <Reaction SGAT_p at 0x7fa537ceff10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT2G19900 or AT5G11670 or AT5G25880)' for <Reaction MalDH2NADP_c at 0x7fa537c99450>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT5G11670 or AT5G25880 or AT1G79750)' for <Reaction MalDH2NADP_h at 0x7fa537c99610>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G10670 or AT1G60810 or AT1G09430) and 4*(AT3G06650 orAT5G49460)' for <Reaction ATPCitL_c at 0x7fa537cb01d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT5G16390 or AT5G15530) and AT5G35360 and 2*(AT2G38040) and 2*(ATCG00500)' for <Reaction ACoAC_h at 0x7fa537cb0850>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G35460 or AT2G17420 or AT2G41680)' for <Reaction TRXR_c at 0x7fa537cb0dd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G12440 or AT5G11160 or AT4G22570 or AT1G27450 or AT1G80050)' for <Reaction ADPRPPT_c at 0x7fa537cb0f50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G77760 or AT1G37130)' for <Reaction NO3R1_c at 0x7fa537cc4710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G77760 or AT1G37130)' for <Reaction NO3R2_c at 0x7fa537cc4550>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G22890 or AT1G19920 or AT4G14680 or AT5G43780)' for <Reaction ATPSL_h at 0x7fa537cc4a10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G54660)' for <Reaction GSSGR_h at 0x7fa537cc4bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT1G08065 or AT5G04180 or AT4G20990 or AT4G21000 or AT2G28210 or AT3G52720) or 8*(AT1G23730 or AT5G14740)' for <Reaction HCO3DHA_c at 0x7fa537cc4d10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '(AT4G20990 or AT1G08080 or AT3G52720) or 8*(AT1G23730 or AT1G70410 or AT1G58180 or AT3G01500 or AT5G14740 or AT4G33580)' for <Reaction HCO3DHA_h at 0x7fa537cc4850>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT3G57560)' for <Reaction AGluK at 0x7fa537cc4dd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT5G08100) and 2*(AT3G16150)' for <Reaction Asnase_c at 0x7fa537c5acd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT2G17640 or AT3G13110 or AT4G35640 or AT5G56760)' for <Reaction SerAcT_c at 0x7fa537c5af10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT3G13110)' for <Reaction SerAcT_h at 0x7fa537c5aed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT1G55920 or AT3G13110)' for <Reaction SerAcT_m at 0x7fa537c6b390>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G14880 or AT3G22460)' for <Reaction CysS_c at 0x7fa537c6b3d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G61440 or AT3G03630 or AT4G14880 or AT2G43750)' for <Reaction CysS_h at 0x7fa537c6b650>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G61440 or AT3G03630 or AT2G43750 or AT5G28020)' for <Reaction CysS_m at 0x7fa537c5aad0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440 or AT3G03910)' for <Reaction GluDH1NAD_m at 0x7fa537c6b1d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440 or AT3G03910)' for <Reaction GluDH1NADP_m at 0x7fa537c6bb50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G18170 or AT5G07440)' for <Reaction GluDH2NAD_m at 0x7fa537c6be10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT5G17330 or AT1G65960 or AT2G02000 or AT2G02010 or AT3G17760)' for <Reaction GluDC_c at 0x7fa537c6bed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G22200)' for <Reaction GABATA1_m at 0x7fa537c6bb90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G22200)' for <Reaction GABATA2_m at 0x7fa537c6bd10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G79440)' for <Reaction SCASeADH_m at 0x7fa537c6bd50>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G37600 or AT3G17820 or AT5G16570 or AT1G66200)' for <Reaction GlnS_c at 0x7fa537c6bdd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT3G17820 or AT1G48470 or AT5G35630)' for <Reaction GlnS_h at 0x7fa537c6bbd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '8*(AT5G35630)' for <Reaction GlnS_m at 0x7fa537c855d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G20620 or AT1G20630 or AT4G35090)' for <Reaction CAT_h at 0x7fa537c85050>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G13360' for <Reaction SGAT_h at 0x7fa537c85190>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G10050)' for <Reaction ThrDA_h at 0x7fa537c1c410>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS1_h at 0x7fa537c1c4d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS2_h at 0x7fa537c1c690>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G48560)' for <Reaction ALS3_h at 0x7fa537c1c8d0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G18500) or 4*(AT1G74040)' for <Reaction 2IPMS_h at 0x7fa537c1cd10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G19710 or AT1G31230 or AT5G13280 or AT3G02020 or AT5G14060)' for <Reaction AspK_h at 0x7fa537c1ced0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G14810)' for <Reaction Asp_DASH_SeADH_h at 0x7fa537c1cfd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G33680)' for <Reaction DAPAT_h at 0x7fa537c2f410>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G31230 AT4G19710)' for <Reaction HSerDHNADP_h at 0x7fa537c2f990>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT1G31230 AT4G19710)' for <Reaction HSerDHNAD_h at 0x7fa537c2f550>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT3G57050)' for <Reaction CTHL_h at 0x7fa537c2fdd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G44160 or AT3G59970)' for <Reaction 5M_DASH_THFor_c at 0x7fa537c2fbd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '4*(AT4G13930 or AT4G13890 or AT1G22020 or AT1G36370 or AT4G37930 or AT5G26780)' for <Reaction GlyHMT_c at 0x7fa537c2ffd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT4G29840 or AT1G72810)' for <Reaction ThrS_h at 0x7fa537bd7c10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT1G24807) and 2*(AT3G55870 or AT5G05730 or AT2G29690)' for <Reaction ANTS_h at 0x7fa537bd7910>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G38280)' for <Reaction AMPDA_c at 0x7fa537bebcd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT3G57610)' for <Reaction DCAMPS_h at 0x7fa537bebe10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction ADPR_c at 0x7fa537c00950>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction CDPR_c at 0x7fa537c00b90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction GDPR_c at 0x7fa537c00f10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '2*(AT2G21790) and 2*(AT3G23580 or AT3G27060)' for <Reaction UDPR_c at 0x7fa537c00bd0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '3*(AT3G46940)' for <Reaction dUTPP_c at 0x7fa537c00c90>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction GDPK_c at 0x7fa537c00e10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction GDPK_h at 0x7fa537c00ed0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction UDPK_c at 0x7fa537c13710>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dADPK_c at 0x7fa537c13850>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dCDPK_c at 0x7fa537c13990>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dGDPK_c at 0x7fa537c13ad0>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dTDPK_c at 0x7fa537c13c10>\n",
      "cobra/core/reaction.py:394 \u001b[1;31mUserWarning\u001b[0m: malformed gene_reaction_rule '6*(AT4G09320 or AT5G63310)' for <Reaction dUDPK_c at 0x7fa537c13d50>\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mal_c --> Fum_c + H2O_c\n",
      "(0.0, 1000.0)\n",
      "Directionality changed. Dyson et al. 2016 \n",
      "\n",
      "Mal_c --> \n",
      "Fum_c --> \n",
      "starch1_h --> \n",
      "G6P_c + Pi_h --> G6P_h + Pi_c\n",
      "G6P_c + Pi_h <=> G6P_h + Pi_c\n",
      "(-1000.0, 1000.0)\n",
      "Set to be bidirectional. Dyson et al. 2015 \n",
      "\n",
      "G6P_c + Pi_h <=> G6P_h + Pi_c\n",
      "(0.0, 1000.0)\n",
      "Set as one directional to avoid futile cycles\n",
      "G6P_c + Pi_h --> G6P_h + Pi_c\n",
      "(0.0, 1000.0)\n",
      "Set as one directional to avoid futile cycles\n",
      "G6P_c + Pi_h --> G6P_h + Pi_c\n",
      "(0.0, 1000.0)\n",
      "Set as one directional to avoid futile cycles\n"
     ]
    }
   ],
   "source": [
    "# Loading the model \n",
    "%run ModelSetUp.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions for running the models \n",
    "\n",
    "def ATPNADPH(atp,nadph):\n",
    "    '''Setting the ATP and NADPH flux. Lower bound plus one'''\n",
    "    NADPH = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "    ATPase = model.reactions.get_by_id('ATPase_h')\n",
    "    NADPH.lower_bound = nadph\n",
    "    NADPH.upper_bound = nadph + 1.0\n",
    "    ATPase.lower_bound = atp\n",
    "    ATPase.upper_bound = atp + 1.0\n",
    "    print(NADPH)\n",
    "    print(NADPH.bounds)\n",
    "    print(ATPase)\n",
    "    print(ATPase.bounds)\n",
    "    print(\" \")\n",
    "    \n",
    "def setobjective(FunName,obj=\"Maximum\"):\n",
    "    '''Setting the objective function of the model to FunName'''\n",
    "    model.objective = model.reactions.get_by_id(FunName)\n",
    "    model.objective_direction = obj\n",
    "    print(model.objective)\n",
    "    print(\"Set as objective function. \\n\")\n",
    "    \n",
    "def constr(reacs,Dat1,Dat2,Dat3,Dat4):\n",
    "    '''Constraints model reactions for a specific genotype and temperature'''\n",
    "    for r in reacs:\n",
    "        upper_vals = []\n",
    "        lower_vals = []\n",
    "        for g in r.genes:\n",
    "            # Get index of protein in data file \n",
    "            ind = all_atg.index(g.id)\n",
    "            # Scaling the values for setting fluxes \n",
    "            vals = [i/4000. for i in Dat1[ind],Dat2[ind],Dat3[ind],Dat4[ind]]\n",
    "            val = np.mean(vals)\n",
    "            sterr = np.std(vals)/np.sqrt(len(vals))\n",
    "            upper = val + sterr\n",
    "            lower = val - sterr\n",
    "            upper_vals.append(upper)\n",
    "            lower_vals.append(lower)\n",
    "        # Constraint flux of forward reaction\n",
    "        r.upper_bound = round(sum(upper_vals),3)\n",
    "        # Constraint flux of lower reaction \n",
    "        if r.lower_bound < 0.0:\n",
    "            r.lower_bound = round(-sum(upper_vals),3)\n",
    "        # No lower bound b/c don't know localization or enzymes are used to in other reactions \n",
    "\n",
    "def TPTratio(TPT1,TPT2):\n",
    "    '''To set TPT1 to TPT2 flux ratios'''\n",
    "    r = model.reactions.get_by_id('Tr_TPT2')\n",
    "    r.lower_bound = 0.0\n",
    "    r.upper_bound = TPT2\n",
    "    r = model.reactions.get_by_id('Tr_TPT1')\n",
    "    r.lower_bound = 0.0\n",
    "    r.upper_bound = TPT1\n",
    "    #print(cobra.flux_analysis.flux_variability_analysis(model, [NADPH,ATPase]))\n",
    "    print(\" \")\n",
    "    #inds = (1,101,202,302,401,501,601,701,801,901)\n",
    "    \n",
    "def subsys(subsystem):\n",
    "    for r in model.reactions:\n",
    "        if r.subsystem == subsystem:\n",
    "            print(r.reaction)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maximize\n",
      "-1.0*FumHA_c_reverse_5dd7d + 1.0*FumHA_c\n",
      "Set as objective function. \n",
      "\n",
      "Fd_DASH_NADPR_h: 2.0 Fdrd_h + H_h + NADP_h --> 2.0 Fdox_h + NADPH_h\n",
      "(160.0, 161.0)\n",
      "ATPase_h: 3.0 ADP_h + 14.0 H_l + 3.0 Pi_h --> 3.0 ATP_h + 3.0 H2O_h + 11.0 H_h\n",
      "(107.0, 108.0)\n",
      " \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>maximum</th>\n",
       "      <th>minimum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Tr_TPT1</th>\n",
       "      <td>928.571429</td>\n",
       "      <td>-35.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tr_TPT2</th>\n",
       "      <td>1000.000000</td>\n",
       "      <td>-857.142857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             maximum     minimum\n",
       "Tr_TPT1   928.571429  -35.714286\n",
       "Tr_TPT2  1000.000000 -857.142857"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Setting the ATP and NADPH production fluxes \n",
    "setobjective(\"FumHA_c\")\n",
    "ATPNADPH(107.0,160.0)\n",
    "flux_variability_analysis(model, [\"Tr_TPT1\",\"Tr_TPT2\"])\n",
    "#subsys(\"light reactions\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total measured proteins: 2427\n",
      "Total measured proteins in model: 245\n",
      "Total model reactions: 552\n",
      " \n",
      "Total model reactions with single constraints: 62\n",
      "Model proteins for single constraints: 49\n",
      " \n",
      "Total model reactions with multiple constraints: 39\n",
      "Total model proteins for possible constraints: 53\n"
     ]
    }
   ],
   "source": [
    "# Get list of all possible proteins in the model\n",
    "gene_ids = []\n",
    "for g in model.genes:\n",
    "    gene_ids.append(g.id)\n",
    "# Get list of all measured proteins \n",
    "all_atg = []\n",
    "x=df[\"Accession\"]\n",
    "for atg in x.values:\n",
    "    all_atg.append(atg)\n",
    "# Comparing the measured protein concentrations to those which are in the model \n",
    "model_atg = [] # Total measured proteins in model\n",
    "model_atg_final = [] # Proteins with direct reaction constraint (i.e. single responsible protein)\n",
    "model_rxns_atg = [] # Model reactions with direct constraints (i.e. single responsible protein)\n",
    "model_rxns_atgs = [] # Model reactions with multiple possible constraints \n",
    "model_atgs_final = [] # Proteins with multiple possible constraints \n",
    "\n",
    "for atg in x.values: # Check if protein is in the model \n",
    "    if atg in gene_ids:\n",
    "        model_atg.append(atg)\n",
    "        # Only parameterize that reaction if that enzyme is the only one responsible for that reaction\n",
    "        for r in model.genes.get_by_id(atg).reactions:\n",
    "            if len(r.genes) == 1:\n",
    "                model_rxns_atg.append(r)\n",
    "                model_atg_final.append(atg)\n",
    "            else:\n",
    "                prot_names = []\n",
    "                for g in r.genes:\n",
    "                    prot_names.append(g.id)\n",
    "                if set(prot_names).issubset(set(all_atg)):\n",
    "                    model_rxns_atgs.append(r)\n",
    "                    model_atgs_final = model_atgs_final + prot_names\n",
    "# Feasible constraint summary \n",
    "print(\"Total measured proteins: {}\".format(len(all_atg)))  \n",
    "print(\"Total measured proteins in model: {}\".format(len(model_atg))) \n",
    "print(\"Total model reactions: {}\".format(len(model.reactions)))\n",
    "print(\" \")\n",
    "# Reactions with simple constraints \n",
    "print(\"Total model reactions with single constraints: {}\".format(len(set(model_rxns_atg))))\n",
    "print(\"Model proteins for single constraints: {}\".format(len(set(model_atg_final))))\n",
    "print(\" \")\n",
    "# Reactions with multiple possible constraints \n",
    "print(\"Total model reactions with multiple constraints: {}\".format(len(set(model_rxns_atgs))))\n",
    "print(\"Total model proteins for possible constraints: {}\".format(len(set(model_atgs_final))))\n",
    "# Total constained reactions\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs) #these are the reactions constrained in the models \n",
    "constr_ids = []\n",
    "for r in constr_reacs:\n",
    "    constr_ids.append(r.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setting up the Model Constraints\n",
    "# for different genotypes and temperatures \n",
    "\n",
    "def all_constr(C,M,F,S,Cond):\n",
    "    # Metabolite Constraints \n",
    "    f1=0.99*200\n",
    "    f2=1.01*200\n",
    "    rMal = model.reactions.get_by_id('Mal_Store')\n",
    "    rFum = model.reactions.get_by_id('Fum_Store')\n",
    "    rStarch = model.reactions.get_by_id('Starch_Store')\n",
    "    rCO2 = model.reactions.get_by_id('Im_CO2')\n",
    "    FumHA_c = model.reactions.get_by_id('FumHA_c')\n",
    "    rCO2.upper_bound = C+1.0\n",
    "    rCO2.lower_bound = C-1.0\n",
    "    rResp.lower_bound = rCO2.lower_bound*0.2/3\n",
    "    rResp.upper_bound = 1000.0\n",
    "    rMal.upper_bound = (M+0.001)*f2\n",
    "    rMal.lower_bound = (M-0.001)*f1\n",
    "    rFum.upper_bound = (F+0.001)*f2\n",
    "    rFum.lower_bound = (F-0.001)*f1\n",
    "    rStarch.upper_bound = (S-0.001)*f2\n",
    "    rStarch.lower_bound = (S-0.001)*f1\n",
    "    for r in [rCO2,rResp,rMal,rFum,rStarch]:\n",
    "        print(r)\n",
    "        print(r.bounds)\n",
    "        print(\" \")\n",
    "    # Proteomic Constraints \n",
    "    name1 = \"{} 1\".format(Cond)\n",
    "    name2 = \"{} 2\".format(Cond)\n",
    "    name3 = \"{} 3\".format(Cond)\n",
    "    name4 = \"{} 4\".format(Cond)\n",
    "    constr(constr_reacs,df[name1],df[name2],df[name3],df[name4])\n",
    "    # Loopless Constraints \n",
    "    print(\"For Loopless...\")\n",
    "    setobjective(\"Im_CO2\",obj=\"Maximum\")\n",
    "    loopless = loopless_solution(model)\n",
    "    print(loopless[\"Tr_TPT1\"])\n",
    "    print(loopless[\"Tr_TPT2\"])\n",
    "    for r in model.reactions:\n",
    "        if (loopless[r.id] == 0.0):\n",
    "            r.lower_bound = 0.0\n",
    "            r.upper_bound = 0.0\n",
    "        else:\n",
    "            pass \n",
    "    FVA = flux_variability_analysis(model,[\"Tr_TPT1\",\"Tr_TPT2\"])\n",
    "    modelname = \"Arnold2014_{}_Loopless.xml\".foramt(Cond)\n",
    "    cobra.io.write_sbml_model(model,modelname,use_fbc_package=False)\n",
    "    print(\"Model saved. Go do flux sampling in MATLAB.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Im_CO2:  --> CO2_c\n",
      "(99.0, 101.0)\n",
      " \n",
      "PyrK_c: ADP_c + H_c + PEP_c --> ATP_c + Pyr_c\n",
      "(6.6000000000000005, 1000.0)\n",
      " \n",
      "Mal_Store: Mal_c --> \n",
      "(0.396, 0.808)\n",
      " \n",
      "Fum_Store: Fum_c --> \n",
      "(0.99, 1.414)\n",
      " \n",
      "Starch_Store: starch1_h --> \n",
      "(5.346, 5.454)\n",
      " \n",
      "For Loopless...\n",
      "Maximize\n",
      "-1.0*Im_CO2_reverse_a674e + 1.0*Im_CO2\n",
      "Set as objective function. \n",
      "\n",
      "17.64515929203541\n",
      "2.961513274336143\n"
     ]
    }
   ],
   "source": [
    "# Parameterizing the model for Col-0 Control Conditions\n",
    "all_constr(100,0.003,0.006,0.028,\"WT Cntl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'model_rxns_atg' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-d5df8552a51d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;31m# Set up reactions which will be constrained\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mconstr_reacs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel_rxns_atg\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmodel_rxns_atgs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;31m# Constraining the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'model_rxns_atg' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Col0 Control\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rResp.lower_bound = rCO2.lower_bound*0.2/3\n",
    "rResp.upper_bound = 1000.0\n",
    "rMal.upper_bound = 0.004*f2\n",
    "rMal.lower_bound = 0.002*f1\n",
    "rFum.upper_bound = 0.007*f2\n",
    "rFum.lower_bound = 0.005*f1\n",
    "rStarch.upper_bound = 0.029*f2\n",
    "rStarch.lower_bound = 0.027*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "constr(constr_reacs,df[\"WT Cntl 1\"],df[\"WT Cntl 2\"],df[\"WT Cntl 3\"],df[\"WT Cntl 4\"])\n",
    "\n",
    "# Loopless FBA\n",
    "model.objective = model.reactions.get_by_id(\"FumHA_c\")\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "\n",
    "loopless = loopless_solution(model)\n",
    "print(loopless[\"Tr_TPT1\"])\n",
    "print(loopless[\"Tr_TPT2\"])\n",
    "\n",
    "for r in model.reactions:\n",
    "    if (loopless[r.id] == 0.0):\n",
    "        r.lower_bound = 0.0\n",
    "        r.upper_bound = 0.0\n",
    "    else:\n",
    "        pass \n",
    "\n",
    "FVA = flux_variability_analysis(model, model.reactions)\n",
    "print(FVA.loc[\"FumHA_c\"])\n",
    "print(FVA.loc[\"Tr_TPT1\"])\n",
    "print(FVA.loc[\"Tr_TPT2\"])\n",
    "\n",
    "# Saving the model\n",
    "# cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cntl_Loopless.xml\",use_fbc_package=False)\n",
    "\n",
    "#TPTratio(1000.0,0.0)\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cntl_tpt2_LooplessC.xml\",use_fbc_package=False)\n",
    "FVA = flux_variability_analysis(model, model.reactions)\n",
    "print(FVA.loc[\"FumHA_c\"])\n",
    "# TPTratio(0.0,1000.0)\n",
    "# cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cntl_tpt1_Loopless.xml\",use_fbc_package=False)\n",
    "# FVA = flux_variability_analysis(model, model.reactions)\n",
    "# print(FVA.loc[\"FumHA_c\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Fum2 Control Conditions\n",
    "\n",
    "# Fum2 Control\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rResp.lower_bound = rCO2.lower_bound*0.2/3\n",
    "rResp.upper_bound = 1000.0\n",
    "rMal.upper_bound = 0.013*f2\n",
    "rMal.lower_bound = 0.011*f1\n",
    "rFum.upper_bound = 0.0*f2\n",
    "rFum.lower_bound = 0.0*f1\n",
    "rStarch.upper_bound = 0.043*f2\n",
    "rStarch.lower_bound = 0.041*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "r = model.reactions.get_by_id('FumHA_c')\n",
    "r.upper_bound = 0.0 # Mutant knockout \n",
    "r.lower_bound = 0.0 \n",
    "constr(constr_reacs,df[\"fum2 Cntl 1\"],df[\"fum2 Cntl 2\"],df[\"fum2 Cntl 3\"],df[\"fum2 Cntl 4\"]) #Proteomics \n",
    "\n",
    "# Loopless FBA\n",
    "model.objective = model.reactions.get_by_id(\"FumHA_c\")\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "\n",
    "loopless = loopless_solution(model)\n",
    "print(loopless[\"Tr_TPT1\"])\n",
    "print(loopless[\"Tr_TPT2\"])\n",
    "\n",
    "for r in model.reactions:\n",
    "    if (loopless[r.id] == 0.0):\n",
    "        r.lower_bound = 0.0\n",
    "        r.upper_bound = 0.0\n",
    "    else:\n",
    "        pass \n",
    "\n",
    "FVA = flux_variability_analysis(model, model.reactions)\n",
    "print(FVA.loc[\"FumHA_c\"])\n",
    "print(FVA.loc[\"Tr_TPT1\"])\n",
    "print(FVA.loc[\"Tr_TPT2\"])\n",
    "\n",
    "# # Saving the model\n",
    "# cobra.io.write_sbml_model(model,\"Arnold2014_M_Cntl.xml\",use_fbc_package=False)\n",
    "\n",
    "# TPTratio(1000.0,0.0)\n",
    "cobra.io.write_sbml_model(model,\"Arnold2014_M_Cntl_tpt1_LooplessC.xml\",use_fbc_package=False)\n",
    "# TPTratio(0.0,1000.0)\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_M_Cntl_tpt1_Loopless.xml\",use_fbc_package=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Col-0 Cold Conditions\n",
    "\n",
    "# Col0 Cold\n",
    "rCO2.upper_bound = 101.0\n",
    "rCO2.lower_bound = 99.0\n",
    "rResp.lower_bound = rCO2.lower_bound*0.2/3\n",
    "rResp.upper_bound = 1000.0\n",
    "rMal.upper_bound = 0.013*f2\n",
    "rMal.lower_bound = 0.011*f1\n",
    "rFum.upper_bound = 0.024*f2\n",
    "rFum.lower_bound = 0.022*f1\n",
    "rStarch.upper_bound = 0.052*f2\n",
    "rStarch.lower_bound = 0.05*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "constr(constr_reacs,df[\"WT cold 1\"],df[\"WT cold 2\"],df[\"WT cold 3\"],df[\"WT cold 4\"])\n",
    "\n",
    "# Loopless FBA\n",
    "model.objective = model.reactions.get_by_id(\"FumHA_c\")\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "\n",
    "loopless = loopless_solution(model)\n",
    "print(loopless[\"Tr_TPT1\"])\n",
    "print(loopless[\"Tr_TPT2\"])\n",
    "\n",
    "for r in model.reactions:\n",
    "    if (loopless[r.id] == 0.0):\n",
    "        r.lower_bound = 0.0\n",
    "        r.upper_bound = 0.0\n",
    "    else:\n",
    "        pass \n",
    "\n",
    "FVA = flux_variability_analysis(model, model.reactions)\n",
    "print(FVA.loc[\"FumHA_c\"])\n",
    "print(FVA.loc[\"Tr_TPT1\"])\n",
    "print(FVA.loc[\"Tr_TPT2\"])\n",
    "\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cold.xml\",use_fbc_package=False)\n",
    "\n",
    "#TPTratio(1000.0,0.0)\n",
    "cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cold_tpt2_LooplessC.xml\",use_fbc_package=False)\n",
    "#TPTratio(0.0,1000.0)\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_WT_Cold_tpt1.xml\",use_fbc_package=False)\n",
    "\n",
    "# for i in range(len(inds)):\n",
    "#     TPTratio(inds[i],inds[-(i+1)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameterizing the model for Fum2 Cold Conditions\n",
    "\n",
    "# Fum2 Cold\n",
    "rCO2.upper_bound = 95.0\n",
    "rCO2.lower_bound = 93.0\n",
    "rResp.lower_bound = rCO2.lower_bound*0.2/3\n",
    "rResp.upper_bound = 1000.0\n",
    "rMal.upper_bound = 0.026*f2\n",
    "rMal.lower_bound = 0.024*f1\n",
    "rFum.upper_bound = 0.0*f2\n",
    "rFum.lower_bound = 0.0*f1\n",
    "rStarch.upper_bound = 0.064*f2\n",
    "rStarch.lower_bound = 0.062*f1\n",
    "\n",
    "# Set up reactions which will be constrained\n",
    "constr_reacs = set(model_rxns_atg + model_rxns_atgs)\n",
    "\n",
    "# Constraining the model\n",
    "r = model.reactions.get_by_id('FumHA_c')\n",
    "r.upper_bound = 0.0 # Mutant knockout \n",
    "r.lower_bound = 0.0 \n",
    "constr(constr_reacs,df[\"fum2 Cold 1\"],df[\"fum2 Cold 2\"],df[\"fum2 Cold 3\"],df[\"fum2 Cold 4\"]) #Proteomics \n",
    "\n",
    "# Loopless FBA\n",
    "model.objective = model.reactions.get_by_id(\"FumHA_c\")\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "\n",
    "loopless = loopless_solution(model)\n",
    "print(loopless[\"Tr_TPT1\"])\n",
    "print(loopless[\"Tr_TPT2\"])\n",
    "\n",
    "for r in model.reactions:\n",
    "    if (loopless[r.id] == 0.0):\n",
    "        r.lower_bound = 0.0\n",
    "        r.upper_bound = 0.0\n",
    "    else:\n",
    "        pass \n",
    "\n",
    "FVA = flux_variability_analysis(model, model.reactions)\n",
    "print(FVA.loc[\"FumHA_c\"])\n",
    "print(FVA.loc[\"Tr_TPT1\"])\n",
    "print(FVA.loc[\"Tr_TPT2\"])\n",
    "\n",
    "# Saving the model\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_M_Cold.xml\",use_fbc_package=False)\n",
    "\n",
    "#TPTratio(1000.0,0.0)\n",
    "cobra.io.write_sbml_model(model,\"Arnold2014_M_Cold_tpt2_LooplessC.xml\",use_fbc_package=False)\n",
    "#TPTratio(0.0,1000.0)\n",
    "#cobra.io.write_sbml_model(model,\"Arnold2014_M_Cold_tpt1.xml\",use_fbc_package=False)\n",
    "\n",
    "# for i in range(len(inds)):\n",
    "#     TPTratio(inds[i],inds[-(i+1)])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CHRR flux sampling (MATLAB SCRIPT)\n",
    "# Plotting the overlapping sampling distribution (NEW PYTHON SCRIPT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting model reactions as sets \n",
    "req_reacs = []\n",
    "Nreq_reacs = []\n",
    "\n",
    "# Exluding reactions which cannot be deleted in combination with others\n",
    "# Round 1\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2h\"))\n",
    "# Round 2 \n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_CO2h\"))\n",
    "# Round 3\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_H2Oh\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Im_NO3\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2m\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_H2Oh\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_O2m\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_NO2\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_NTT\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_AAC\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT1\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT2\"))\n",
    "exp_reacs.append(model.reactions.get_by_id(\"Tr_TPT3\"))\n",
    "\n",
    "noAT = []\n",
    "for r in model.reactions:\n",
    "    g = r.genes\n",
    "    for gen in g:\n",
    "        if gen.id == '-':\n",
    "            if r not in exp_reacs:\n",
    "                noAT.append(r)\n",
    "                r.upper_bound = 0.0\n",
    "                r.lower_bound = 0.0\n",
    "                \n",
    "print(len(noAT))\n",
    "\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "# # Trying to delete as high a set of them as possible \n",
    "# for s in set(itertools.combinations(noAT, 4)): #set(itertools.combinations(noAT, 1))\n",
    "#     print(\"Check\")\n",
    "#     b = []\n",
    "#     for i in range(len(s)):\n",
    "#         b.append(s[i].bounds)\n",
    "#         s[i].upper_bound = 0.0\n",
    "#         s[i].lower_bound = 0.0\n",
    "#     interest_reacs = [\"Im_CO2\",\"Mal_Store\",\"Fum_Store\",\"Starch_Store\",\"Ex_Suc\"]\n",
    "#     test = []\n",
    "#     for r in interest_reacs:\n",
    "#         model.objective = r\n",
    "#         try:\n",
    "#             bounds = flux_variability_analysis(model, r)\n",
    "#             if bounds['minimum'][0] > 0.0:\n",
    "#                 test.append(0) # Carbon storage ok\n",
    "#             else:\n",
    "#                 test.append(1)\n",
    "#         except Exception:\n",
    "#             test.append(1)\n",
    "#     if sum(test) > 0: # Can't store carbon \n",
    "#         req_reacs.append(s)\n",
    "#     else:\n",
    "#         Nreq_reacs.append(s)\n",
    "#     for i in range(len(s)):\n",
    "#         # Resetting the reactions bounds\n",
    "#         s[i].upper_bound = b[i][1]\n",
    "#         s[i].lower_bound = b[i][0]\n",
    "# print(\"Finished\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Deleting model reactions as sets \n",
    "req_reacs = []\n",
    "Nreq_reacs = []\n",
    "\n",
    "# Trying to delete as high a set of them as possible \n",
    "for s in set(itertools.combinations(model.reactions, 1)): #set(itertools.combinations(noAT, 1))\n",
    "    b = []\n",
    "    for i in range(len(s)):\n",
    "        b.append(s[i].bounds)\n",
    "        s[i].upper_bound = 0.0\n",
    "        s[i].lower_bound = 0.0\n",
    "    interest_reacs = [\"Im_CO2\",\"Mal_Store\",\"Starch_Store\",\"Ex_Suc\"]\n",
    "    test = []\n",
    "    for r in interest_reacs:\n",
    "        model.objective = r\n",
    "        try:\n",
    "            bounds = flux_variability_analysis(model, r)\n",
    "            if bounds['minimum'][0] > 0.0:\n",
    "                test.append(0) # Carbon storage ok\n",
    "            else:\n",
    "                test.append(1)\n",
    "        except Exception:\n",
    "            test.append(1)\n",
    "    if sum(test) > 0: # Can't store carbon \n",
    "        req_reacs.append(s)\n",
    "    else:\n",
    "        Nreq_reacs.append(s)\n",
    "    for i in range(len(s)):\n",
    "        # Resetting the reactions bounds\n",
    "        s[i].upper_bound = b[i][1]\n",
    "        s[i].lower_bound = b[i][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rObj = model.reactions.get_by_id('Im_CO2')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "rObj = model.reactions.get_by_id('ATPase_h')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Minimum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))\n",
    "\n",
    "rObj = model.reactions.get_by_id('Fd_DASH_NADPR_h')\n",
    "print(rObj.reaction)\n",
    "model.objective = rObj\n",
    "model.objective_direction = \"Maximum\"\n",
    "print(model.objective)\n",
    "print(cobra.flux_analysis.flux_variability_analysis(model, all_reac))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Appending transport reactions which are not to be deleted\n",
    "r1 = model.reactions.get_by_id(\"Tr_TPT1\")\n",
    "r2 = model.reactions.get_by_id(\"Tr_TPT2\")\n",
    "\n",
    "r = r1\n",
    "r.upper_bound = 0.0\n",
    "r.lower_bound = 0.0\n",
    "\n",
    "\n",
    "# Setting reactions without AT codes to zero\n",
    "noAT = []\n",
    "for r in model.reactions:\n",
    "    g = r.genes\n",
    "    for gen in g:\n",
    "        if gen.id == '-':\n",
    "            if r not in exp_reacs:\n",
    "#                 if r.id[0] is not \"T\":\n",
    "#                     print(r.id)\n",
    "#                     print(r.reaction)\n",
    "#                     r.upper_bound = 0.0\n",
    "#                     r.lower_bound = 0.0\n",
    "                noAT.append(r)\n",
    "print(len(noAT))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
